### PR TITLE
stb_truetype memory management through the nk_font_atlas allocator

### PR DESCRIPTION
**Pull request by 2asoft**
After the tremendous work done to [update stb_truetype](https://github.com/Immediate-Mode-UI/Nuklear/pull/53) to a newer version, Nuklear lost the ability to control stb_truetype's memory management via Nuklear's allocators, instead delegating it to the default (`malloc`/`free`) allocation scheme.
In cases where Nuklear is compiled without the default allocation scheme, the entire library would fail to build.
More details in [this comment](https://github.com/Immediate-Mode-UI/Nuklear/pull/53#discussion_r483763574).

This PR implements stb_truetype `STBTT_malloc`/`STBTT_free` to delegate to the `nk_allocator` used by `nk_font_atlas`.
It also leaves the ability for the consumer to define own `STBTT_malloc`/`STBTT_free` implementations if they so desire.

### CODE REVIEW

File: nuklear.h [hunk 1]
```diff
@@ -5684,18 +5684,6 @@ template<typename T> struct nk_alignof{struct Big {T x; char c;}; enum {
 #define NK_ALIGNOF(t) ((char*)(&((struct {char c; t _h;}*)0)->_h) - (char*)0)
 #endif
 
-#ifdef NK_IMPLEMENTATION
```
Comment (2asoft) [inline nuklear.h:5687]: This file is rebuilt with `paq.sh`.

File: src/paq.sh [hunk 2]
```diff
@@ -1,3 +1,3 @@
...
+python build.py --macro NK --intro HEADER --pub nuklear.h --priv1 nuklear_internal.h,nuklear_math.c,nuklear_util.c,nuklear_color.c,nuklear_utf8.c,nuklear_buffer.c,nuklear_string.c,nuklear_draw.c,nuklear_vertex.c --extern stb_rect_pack.h,stb_truetype.h --priv2 nuklear_font.c,nuklear_input.c,nuklear_style.c,nuklear_context.c,nuklear_pool.c,nuklear_page_element.c,nuklear_table.c,nuklear_panel.c,nuklear_window.c,nuklear_popup.c,nuklear_contextual.c,nuklear_menu.c,nuklear_layout.c,nuklear_tree.c,nuklear_group.c,nuklear_list_view.c,nuklear_widget.c,nuklear_text.c,nuklear_image.c,nuklear_button.c,nuklear_toggle.c,nuklear_selectable.c,nuklear_slider.c,nuklear_progress.c,nuklear_scrollbar.c,nuklear_text_editor.c,nuklear_edit.c,nuklear_property.c,nuklear_chart.c,nuklear_color_picker.c,nuklear_combo.c,nuklear_tooltip.c --outro LICENSE,CHANGELOG,CREDITS > ../nuklear.h
```
Comment (2asoft) [inline src/paq.sh:2]: Added redirect to prebuilt `nuklear.h` file, as in `paq.bat`.

File: src/nuklear.h [hunk 3]
```diff
@@ -5465,16 +5465,4 @@ template<typename T> struct nk_alignof{struct Big {T x; char c;}; enum {
 #define NK_ALIGNOF(t) ((char*)(&((struct {char c; t _h;}*)0)->_h) - (char*)0)
 #endif
 
-#ifdef NK_IMPLEMENTATION
```
Comment (2asoft) [inline src/nuklear.h:5468]: Moved from here into `src/nuklear_internal.h`.
These are internal implementation details that do not need to be exposed.

File: src/nuklear_font.c [hunk 4]
```diff
@@ -190,7 +190,9 @@ nk_font_bake_pack(struct nk_font_baker *baker,
     /* setup font baker from temporary memory */
     for (config_iter = config_list; config_iter; config_iter = config_iter->next) {
         it = config_iter;
-        do {if (!stbtt_InitFont(&baker->build[i++].info, (const unsigned char*)it->ttf_blob, 0))
+        struct stbtt_fontinfo *font_info = &baker->build[i++].info;
```
Comment (2asoft) [inline src/nuklear_font.c:193]: `stb_truetype` passes the `userdata` field of `stbtt_fontinfo` as the second argument to `STBTT_malloc`/`STBTT_free`.
Set the `userdata` here to point to the allocator (`alloc`), and use it in the custom implementations of `STBTT_malloc`/`STBTT_free`.

File: src/nuklear_internal.h [hunk 5]
```diff
@@ -319,5 +319,31 @@ NK_LIB void nk_draw_property(struct nk_command_buffer *out, const struct nk_styl
 NK_LIB void nk_do_property(nk_flags *ws, struct nk_command_buffer *out, struct nk_rect property, const char *name, struct nk_property_variant *variant, float inc_per_pixel, char *buffer, int *len, int *state, int *cursor, int *select_begin, int *select_end, const struct nk_style_property *style, enum nk_property_filter filter, struct nk_input *in, const struct nk_user_font *font, struct nk_text_edit *text_edit, enum nk_button_behavior behavior);
 NK_LIB void nk_property(struct nk_context *ctx, const char *name, struct nk_property_variant *variant, float inc_per_pixel, const enum nk_property_filter filter);
 
+#ifdef NK_INCLUDE_FONT_BAKING
```
Comment (2asoft) [inline src/nuklear_internal.h:322]: Moved here from `src/nuklear_internal.h`.
These are internal implementation details that do not need to be exposed.

**Comment by 2asoft** [comment]
@dumblob could you please take a look at this PR?
I've verified locally that it works with the demo code in repo.
Do you think there's any additional verification that needs to be done here?

**Comment by dumblob** [comment]
@2asoft sorry for the delay. Yeah, I've looked at it and technically it LGTM. Please take a look at https://github.com/Immediate-Mode-UI/Nuklear#reviewers-guide to add some more bits to have this PR accepted :wink: (e.g. version bump), thanks!

**Comment by 2asoft** [comment]
Thanks. I updated the version, and as far as I can tell all the recommendations in the guide are covered.

**Comment by dumblob** [comment]
Thanks for patience and great feedback!

Btw. would you want to get commit permissions to the repo?

It's basically only about reviews (essentially the things from the Reviewer's guide). As you see, there are only very few reviewers currently and the library development is very slow which is a pity, but we can't afford more time.

**Comment by 2asoft** [comment]
I'll gladly take the commit bit, but I can't devote a large amount of time to Nuklear at the moment.
It does align with some of my current interests so I will be able to spend some time on it, but things may change in the future.

**Comment by dumblob** [comment]
@2asoft invitation sent out :wink:.

Just follow the reviewers guidelines (especially the point "at least two reviewers (both different from the PR author) shall approve `nuklear.h` changes and at least one shall approve changes to other files". That's it. And take your time.

**Comment by dumblob** [comment]
Btw. as you noticed we sometimes break these guidelines, despite we always try hard to stick with them (otherwise issues like https://github.com/Immediate-Mode-UI/Nuklear/issues/198 tend to appear).

**Comment by 2asoft** [comment]
> @2asoft invitation sent out wink.

I didn't receive anything -- can you send it out again please?

**Comment by dumblob** [comment]
@2asoft I'm glad you managed - so welcome to our organization :wink:!

**Pull request merged**